### PR TITLE
unplug_memory: include virtio-mem unplug support

### PR DESCRIPTION
**Pull request by mcasquer**
unplug_memory: include virtio-mem unplug support

Following the same approach implemented in virtio-mem
hotplug; includes the support for virtio-mem device
unplug.

Signed-off-by: mcasquer <mcasquer@redhat.com>
ID: 1939

### CODE REVIEW

**Comment by mcasquer** [comment]
@luckyh @YongxueHong @PaulYuuu please could you review this PR? Thanks!

File: virttest/utils_test/qemu/__init__.py [hunk 1]
```diff
@@ -665,19 +665,22 @@ def unplug_memory(self, vm, name):
         """
         devices = []
         qid_dimm = "dimm-%s" % name
+        qid_virtio_mem = "virtio_mem-%s" % name
```
Comment (PaulYuuu) [inline virttest/utils_test/qemu/__init__.py:668]: Also need to update in doc(L660)

File: virttest/utils_test/qemu/__init__.py [hunk 2]
```diff
@@ -665,19 +665,22 @@ def unplug_memory(self, vm, name):
...
-            devices.append(dimm)
-            self.update_vm_after_unplug(vm, dimm)
+            self.device_unplug(dimm, qid_dimm, devices, vm)
+
+        try:
+            vmem = vm.devices.get_by_qid(qid_virtio_mem)[0]
+        except IndexError:
+            LOG.warn("'%s' is not used by any virtio-mem" % qid_mem)
+        else:
+            self.device_unplug(vmem, qid_virtio_mem, devices, vm)
```
Comment (PaulYuuu) [inline virttest/utils_test/qemu/__init__.py:683]: I think dimm and virtio-mem can in the one `try` code block and use `for` loop, but currently we only 2 types of devices, so not an urgent thing, up to you.

File: virttest/utils_test/qemu/__init__.py (hunk 1 above)
Comment (mcasquer) [reply virttest/utils_test/qemu/__init__.py]: Updated, thanks !

File: virttest/utils_test/qemu/__init__.py (hunk 2 above)
Comment (mcasquer) [reply virttest/utils_test/qemu/__init__.py]: Updated, thanks !

**Comment by mcasquer** [comment]
And it doesn't affect the behavior of dimm unplug cases
```
 (1/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hotplug_mem_share_discard_data.backend_ram.q35: STARTED
 (1/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hotplug_mem_share_discard_data.backend_ram.q35: PASS (92.94 s)
 (2/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hotplug_mem_share_discard_data.backend_file.discard_data_on.q35: STARTED
 (2/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hotplug_mem_share_discard_data.backend_file.discard_data_on.q35: PASS (96.50 s)
 (3/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hotplug_mem_share_discard_data.backend_file.discard_data_off.q35: STARTED
 (3/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hotplug_mem_share_discard_data.backend_file.discard_data_off.q35: PASS (88.12 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

File: virttest/utils_test/qemu/__init__.py [hunk 3]
```diff
@@ -657,27 +657,23 @@ def hotplug_memory(self, vm, name):
...
-        step 1, unplug dimm device
+        step 1, unplug the associated dimm or virtio-mem device
         step 2, unplug memory object
 
         :param vm: VM object
         :param name: memory device name
         """
         devices = []
-        qid_dimm = "dimm-%s" % name
+        qid_list = ["dimm-%s" % name, "virtio_mem-%s" % name]
```
Comment (PaulYuuu) [inline virttest/utils_test/qemu/__init__.py:667]: I have a question: will dimm and virtio-mem use the same `name` in the case? And can they be used together? If no, means here will always log a warning message, right?

Comment (mcasquer) [reply virttest/utils_test/qemu/__init__.py:667]: Yes, there will always be a warning in the logs. Even if you try to use the same name twice in the `target_mems`, for virtio-mem is needed to specify the `vm_memdev_model`, that will lead to a duplicate ID.

File: virttest/utils_test/qemu/__init__.py [hunk 4]
```diff
@@ -657,27 +657,23 @@ def hotplug_memory(self, vm, name):
...
-            devices.append(dimm)
-            self.update_vm_after_unplug(vm, dimm)
+
+        for qid in qid_list:
+            try:
+                device = vm.devices.get_by_qid(qid)[0]
+            except IndexError:
+                LOG.warn("'%s' is not used by %s" % (qid_mem, qid))
+            else:
+                self.device_unplug(device, qid, devices, vm)
```
Comment (luckyh) [inline virttest/utils_test/qemu/__init__.py:676]:
```suggestion
        for qid in qid_list:
            try:
                device = vm.devices.get_by_qid(qid)[0]
            except IndexError:
                pass
            else:
                self.device_unplug(device, qid, devices, vm)
                break
        else:
            LOG.warn("'%s' is not used any memory devices", qid_mem)
```

as

1. need to end the loop once there was any target dev node been found and unplugged
2. report the warning (or raise exceptions instead) only if the given memory object not associated with any memory device

Does that make sense?

Comment (mcasquer) [reply virttest/utils_test/qemu/__init__.py:676]: @luckyh yes, good suggestion ! Thanks :grinning:

Comment (mcasquer) [reply virttest/utils_test/qemu/__init__.py]: Code updated !

File: virttest/utils_test/qemu/__init__.py (hunk 3 above)
Comment (mcasquer) [reply virttest/utils_test/qemu/__init__.py:667]: No warning message anymore after applying Xu's suggestion

**Comment by mcasquer** [comment]
```
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hotplug_virtio_mem.q35: STARTED
9 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.hotplug_virtio_mem.q35: PASS (252.49 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

File: virttest/utils_test/qemu/__init__.py [hunk 5]
```diff
@@ -694,6 +690,22 @@ def unplug_memory(self, vm, name):
         self.update_vm_after_unplug(vm, mem)
         return devices
 
+    @error_context.context_aware
+    def device_unplug(self, device, device_id, devices, vm):
+        """"
```
[the remark this answers is not in the document]
Comment (mcasquer) [reply virttest/utils_test/qemu/__init__.py]: Code updated !

**Review by YongxueHong — APPROVED** [review]

**Review by luckyh — APPROVED** [review]

**Review by PaulYuuu — APPROVED** [review]

**Comment by mcasquer** [comment]
Everybody has approved the PR, could it be merged? Thanks !

**Pull request merged**